### PR TITLE
STM32: Add support for STM32L5 series and nucleo_l552ze_q board

### PR DESCRIPTION
**Pull request by erwango**
This work add support for first STM32 Cortex-M33 based Series.
Support for board nucleo_l552_q is also added with minimum peripheral support for now.

### CODE REVIEW

File: dts/bindings/flash_controller/st,stm32l5-flash-controller.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+description: STM32 L5 flash controller
```
Comment (MaureenHelm) [inline dts/bindings/flash_controller/st,stm32l5-flash-controller.yaml]: Need a license/copyright header

File: boards/arm/nucleo_l552ze_q/doc/nucleol552ze_q.rst [hunk 2]
```diff
@@ -0,0 +1,266 @@
+.. _nucleo_l552ze_q_board:
+
+ST Nucleo L552ZE Q
+##################
+
+Overview
+********
+
+The Nucleo L552ZE Q board features an ARM Cortex-M4 based STM32L552ZE MCU
```
Comment (MaureenHelm) [inline boards/arm/nucleo_l552ze_q/doc/nucleol552ze_q.rst]: M33?

File: boards/arm/nucleo_l552ze_q/doc/nucleol552ze_q.rst [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+More information about the board can be found at the `Nucleo L552ZE Q website`_.
+
+Hardware
+********
+
+The STM32L552ZE SoC provides the following hardware IPs:
+
+- Ultra-low-power with FlexPowerControl (down to 130 nA Standby mode and 100 uA/MHz run mode)
+- Core: ARM |reg| 32-bit Cortex |reg|-M4 CPU with FPU, frequency up to 80 MHz, 100DMIPS/1.25DMIPS/MHz (Dhrystone 2.1)
```
Comment (MaureenHelm) [inline boards/arm/nucleo_l552ze_q/doc/nucleol552ze_q.rst]: M33?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: dts/arm/st/l5/stm32l5.dtsi [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+
+/ {
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		cpu@0 {
+			device_type = "cpu";
+			compatible = "arm,cortex-m33";
```
Comment (ioannisg) [inline dts/arm/st/l5/stm32l5.dtsi]: @erwango  can you add a node for the number of MPU regions? This is not default:8 for M33, as it is for ARMv7-M.

**Review by ioannisg — DISMISSED** [review]
What is the security domain of the added board definition? Looks like the SoC has TrustZone-M so the options are
- default (implying Secure domain)
- non-secure (in that case you need to set CONFIG_TRUSTED_EXECUTION_NONSECURE)

File: dts/arm/st/l5/stm32l5.dtsi (hunk 4 above)
Comment (erwango) [reply dts/arm/st/l5/stm32l5.dtsi]: It's 16 indeed, but split into 8 for secure and 8 for non secure.
So, since we define board as non-secure, it has access to 8.

**Comment by erwango** [comment]
@ioannisg , @MaureenHelm, thanks for the reviews and guidance. I've updated the changes.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Looks fine except for the `west.yml` update

File: west.yml [hunk 5]
```diff
@@ -68,7 +68,7 @@ manifest:
       revision: fa481784b3c49780f18d50bafe00390ccb62b2ec
       path: modules/hal/st
     - name: hal_stm32
-      revision: 37dcc0e120bb2bb99d3ee3b99dbbcf2bffe50258
+      revision: pull/48/head
```
[the remark this answers is not in the document]
Comment (erwango) [reply west.yml]: @MaureenHelm , Done

**Review by MaureenHelm — APPROVED** [review]

File: boards/arm/nucleo_l552ze_q/nucleo_l552ze_q_defconfig [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+# console
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+
+# Enable MPU
+CONFIG_ARM_MPU=y
+
+# This Board implies building Non-Secure firmware
+CONFIG_TRUSTED_EXECUTION_NONSECURE=y
```
Comment (ioannisg) [inline boards/arm/nucleo_l552ze_q/nucleo_l552ze_q_defconfig:45]: @erwango I wonder about this: if you set _NONSECURE by default, one cannot use this platform as standalone in Zephyr, e.g. to run kernel tests, CI samples, etc.: that's because a non-secure firmware requires a secure boot.
Similar V8-M platforms in Zephyr by other vendors have introduced both secure and non-secure target definitions.

I wonder if the STM32 should also align with this.

Comment (erwango) [reply boards/arm/nucleo_l552ze_q/nucleo_l552ze_q_defconfig:45]: @ioannisg by default, TZ is disabled so, we can run zephyr samples on the platform.
I push secure target definitions (basically '_s' version of dts), and we'll be able to run zephyr in secure mode with TZEN=1.
Though, as I understood, in order to be used with TFM (platform upstream is on going on TFM project), tosay zephyr should be run on non secure side. So I didn't planned to upstream secure target definition in zephyr.

Comment (ioannisg) [reply boards/arm/nucleo_l552ze_q/nucleo_l552ze_q_defconfig:45]: You are defining here `CONFIG_TRUSTED_EXECUTION_NONSECURE=y`.
This implies that the Zephyr image is meant to execute in the Non-Secure domain.
This Kconfig is used in a few places in CMake and inside the code where we need to have awareness that the execution domain is non-secure.
Therefore, trying to flash this Nucleo board with the generated image as a _standalone, secure-by-default_ image may lead to problems, e.g. undefined behavior. 

My suggestion is to follow the practice of other TrustZone-enabled platforms in the tree, that is, define two Board definitions, one default, which will be used to run, for example, samples and tests, and one _ns_ target that is meant to run in non-secure mode, side-by-side with some secure firmware, e.g. TF-M. This is done for ARM, Nordic and NXP vendor platforms.

If you don't wish to follow this practice, and want the introduced board to be a strictly non-secure image, that is also fine, but I suggest you state that clearly in the documentation of the board. E.g. saying that flashing the image as-is into an empty Nucleo board does not guarantee to work properly with the zephyr samples' and tests' suite.

Comment (erwango) [reply boards/arm/nucleo_l552ze_q/nucleo_l552ze_q_defconfig:45]: > My suggestion is to follow the practice of other TrustZone-enabled platforms in the tree, that is, define two Board definitions, one default, which will be used to run, for example, samples and tests, and one ns target that is meant to run in non-secure mode, side-by-side with some secure firmware, e.g. TF-M. This is done for ARM, Nordic and NXP vendor platforms.

IIUC, proposed strategy:
- defines a secure target config, that will be used for tests but not used in applications
- defines a non-secure target, that will be used in applications, but not tested

> E.g. saying that flashing the image as-is into an empty Nucleo board does not guarantee to work properly with the zephyr samples' and tests' suite.

Why wouldn't it be possible to consider TZEN=0 as a valid configuration that allows to run zephyr samples & tests?

Comment (ioannisg) [reply boards/arm/nucleo_l552ze_q/nucleo_l552ze_q_defconfig:45]: > defines a secure target config, that will be used for tests but not used in applications

well, it could be used in applications that do not require trust-zone; but, yes, the main point is to have regression testing

>Why wouldn't it be possible to consider TZEN=0 as a valid configuration that allows to run zephyr samples & tests?

I don't understand what TZEN=0 means; is TZEN some macro in the code?

My point is that with NONSECURE_FIRMWARE=y would be bringing build-time configuration different from default. For example, certain parts of CMSE module are not compiled-in, same in fault.c - that means that core ARCH code may behave differently and you might notice differences in tests/samples compared to other v8-m targets.

Btw, have you tried to run the full kernel and arm test suites on this target successfully?

Comment (erwango) [reply boards/arm/nucleo_l552ze_q/nucleo_l552ze_q_defconfig:45]: > I don't understand what TZEN=0 means; is TZEN some macro in the code?

TZEN is an option bit that can be configured by end user. 
Option bits are available to the user to configure hardware options in the controller (like single or dual-bank flash layout or watchdog behavior, but also flash memory protection to configure the mcu as a final product). This is generally done using a tool (STM32CubeProgrammer).

Back to TZEN: it is used to control TrustZone element.
"When the TrustZone is deactivated (TZEN is modified from 1 to 0) after option bytes
loading, the following security features are deactivated:
• Watermark-based secure area
• Block-based secure area
• RDP level 0.5
• Secure interrupt
• All secure registers are RAZ/WI."
By default, on nucleo_l552ze_q, TZEN is set to 0. It is up to the user to modify this and modify TZEN value to enable TrustZone.

You'll find more info in https://www.st.com/resource/en/reference_manual/dm00346336-stm32l552xx-and-stm32l562xx-advanced-armbased-32bit-mcus-stmicroelectronics.pdf

So basically, what I'm pushing is this default out of the box config, with TrustZone disabled.

Regarding kernel tests, 4 are failling:
- tests/kernel/timer/timer_api/kernel.timer.tickless
- tests/kernel/timer/timer_api/kernel.timer
- tests/kernel/mem_protect/userspace/kernel.memory_protection.userspace.gap_filling
- tests/kernel/mem_protect/mem_protect/kernel.memory_protection.gap_filling

I had no time to analyze yet, but I know we have some stability issue on the timer tests, so this may not be M33 specific. I'll have a closer look to the 2 last ones.

Comment (ioannisg) [reply boards/arm/nucleo_l552ze_q/nucleo_l552ze_q_defconfig:45]: I don't think the failures in these tests have much to do with TrustZone

Comment (erwango) [reply boards/arm/nucleo_l552ze_q/nucleo_l552ze_q_defconfig:45]: @ioannisg aside, from the failing tests, do you have concerns in this PR?
Given the information I provided, what would you advise regarding NONSECURE_FIRMWARE?

**Comment by erwango** [comment]
@ioannisg Following discussion held above, (https://github.com/zephyrproject-rtos/zephyr/pull/23551#discussion_r406051580),I removed TRUSTED_EXECUTION_NONSECURE setting.

**Review by FRASTM — COMMENTED** [review]
commit msg 
Add support for stm32l5 socs on stm32 **external** interrupts driver

**Comment by erwango** [comment]
> commit msg
> Add support for stm32l5 socs on stm32 **external** interrupts driver

@FRASTM Fixed, txs

**Review by FRASTM — COMMENTED** [review]
could have a line in the file header too (MP1 seems not here though)
 STM32F0/STM32L0/STM32L4/**STM32L5**/STM32G0/STM32G4: Lines 0 to 15. Lines > 15 are not mapped on an IRQ

**Comment by erwango** [comment]
> could have a line in the file header too (MP1 seems not here though)
STM32F0/STM32L0/STM32L4/STM32L5/STM32G0/STM32G4: Lines 0 to 15. Lines > 15 are not mapped on an IRQ

Done

**Review by FRASTM — APPROVED** [review]

**Comment by ioannisg** [comment]
@erwango I had a look, and removed my -1 on this, to make it go in for v2.3, but, I am confused regarding the objectives here: I see a single board target, now (a single defconfig). This target does not enable any of TRUSTED_EXECUTION_SECURE or _NONSECURE, so when building Zephyr , is this desirable?

I was more thinking you would like to have 2 board targets, here, one "default" to be used for zephyr test suites, and one strictly "non-secure", to build Zephyr as a Non-Secure RTOS (i assumed the latter would need TRUSTED_EXECUTION_NONSECURE=y, but if you managed to set the whole thing up without this need, then I 've got no objections. BTW, the "double board target" is used, no for nRF, Musca, and MPS2 platforms, see respective board definitions ).
